File: checks/python/src/section508/checks/consistent_identification.py
import re

from section508.dom import SPACE, attr, select

GROUPS = [
    [re.compile(r"print", re.I), re.compile(r"download pdf", re.I), re.compile(r"save pdf", re.I), re.compile(r"export pdf", re.I)],
    [re.compile(r"search", re.I), re.compile(r"\bfind\b", re.I)],
    [re.compile(r"log ?in", re.I), re.compile(r"sign in", re.I)],
    [re.compile(r"submit", re.I), re.compile(r"send form", re.I)],
]


def check(root):
    if not root:
        return []

    names = []
    for element in select(root, "button, a[href]"):
        name = SPACE.sub(" ", attr(element, "aria-label") or element.get_text() or "").strip()
        if name:
            names.append(name)

    if len(names) < 2:
        return ["This section has only one named control."]

    for group in GROUPS:
        hits = [name for name in names if any(pattern.search(name) for pattern in group)]
        unique = list(dict.fromkeys(name.lower() for name in hits))
        if len(unique) > 1:
            return [f"“{hits[0]}” and “{hits[1]}” name the same action differently."]
    return []

File: checks/ruby/lib/section508/checks/consistent_identification.rb
# frozen_string_literal: true

module Section508
  module Checks
    # The same function keeps the same name.
    class ConsistentIdentification
      GROUPS = [
        [/print/i, /download pdf/i, /save pdf/i, /export pdf/i],
        [/search/i, /\bfind\b/i],
        [/log ?in/i, /sign in/i],
        [/submit/i, /send form/i]
      ].freeze

      def self.call(root)
        return [] unless root&.respond_to?(:css)

        names = Dom.elements(root, "button, a[href]").map do |element|
          (element["aria-label"] || element.text).to_s.gsub(/\s+/, " ").strip
        end.reject(&:empty?)

        return ["This section has only one named control."] if names.length < 2

        GROUPS.each do |group|
          hits = names.select { |name| group.any? { |pattern| pattern.match?(name) } }
          unique = hits.map(&:downcase).uniq
          if unique.length > 1
            return ["“#{hits[0]}” and “#{hits[1]}” name the same action differently."]
          end
        end
        []
      end
    end
  end
end
